Nickname database

veken1199 opened the pull request:
<!-- You can remove this first section if you have contributed before -->
### First time contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [ ] I have read [how to contribute](https://github.com/WhisperSystems/Signal-Android/blob/master/CONTRIBUTING.md) to this project
- [ ] I have signed the [Contributor License Agreement](https://whispersystems.org/cla/)

### Contributor checklist
<!-- replace the empty checkboxes [ ] below with checked ones [x] accordingly -->
- [ ] I am following the [Code Style Guidelines](https://github.com/WhisperSystems/Signal-Android/wiki/Code-Style-Guidelines)
- [ ] I have tested my contribution on these devices:
 * Device A, Android X.Y.Z
 * Device B, Android Z.Y
 * Virtual device W, Android Y.Y.Z
- [ ] My contribution is fully baked and ready to be merged as is
- [ ] I ensure that all the open issues my contribution fixes are mentioned in the commit message of my first commit using the `Fixes #1234` [syntax](https://help.github.com/articles/closing-issues-via-commit-messages/)

----------

### Description
<!--
Describe briefly what your pull request proposes to fix. Especially if you have more than one commit, it is helpful to give a summary of what your contribution as a whole is trying to solve.
Also, please describe shortly how you tested that your fix actually works.
-->

cool-appz committed:
# related to #34 : Added nick_name attr to database
1- introduced nick_name attribute in both RecipietnDatabase, RecipientSettings
2- introduced nickname modification in DatabaseFactory introducing database helper version 49

cool-appz committed:
related to #38 : Added nick_name attr to database
1- Added RecipientDatabase 'setNickname' public interface
2- Added processNicknameSqlRequest to handle the database calls, this also can be used by removeNickname method in the future
3- Changed the default type of nickname attri in DatabaseFactory from NULL to 'null' - String

cool-appz committed:
related to #38 : Create NicknameHandler - testable:
1- Create class NicknameHandler responsible for setting and unsetting the nickname
2- Sample use:

    NicknameHandler handler = new NicknameHandler(getContext());
    handler.setupDatabaseHandler().setNickname(recipient,"test");
    handler.removeNickname(recipient);

cool-appz committed:
related to #38 : Set Nickname Unit tests

cool-appz committed:
# related to #38 : Set Nickname Unit tests Mocking class

cool-appz committed:
# related to #38 : Set Nickname Unit tests for the mocks
1- also fixed a typo regarding old tests in pin messages
2- moved up the initialization of the mocked object to the parent 'NicknameMocks'

cool-appz committed:
# related to #38 : General code formatting

g-harel commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /**
+     * This method allows the developer to set the underlining database
```
Does this first sentence really say anything that cannot be seen instantly?

veken1199 replied:
Yes, it tells you that This method allows the developer to set the underlining database, it is important to note this point as long as we are using javadoc style

g-harel replied:
It is literally called `SET database`

veken1199 replied:
yes but the rest of the comment is important in my opinion,

veken1199 replied:
yes but this method only used in testing

g-harel replied:
I agree the second part is relevant, that's why I said first sentence in my original comment.

veken1199 replied:
I still believe that line of comment doesn't hurt to keep it. lets see what the other reviewers would say @shelbaz @erinb90 @Dan-Yoo @Radu-Raicea

erinb90 replied:
I think javadoc comments are pretty but I'm still conflicted since the code style guidelines say no comments... I guess keep them for now unless Dr. Shang or Jinfu says otherwise.

g-harel replied:
@erinb90 it isn't about having that javadoc or not. My comment was ONLY about the first sentence.

erinb90 replied:
Oh yeah I agree the first sentence isn't needed.

g-harel commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,70 @@
+     */
+    public boolean setNickname(Recipient recipient, String nickname) {
+        setupDatabaseHandler();
+        return recipientDatabase.setNickname(recipient, nickname);
```
Should this be

```diff
+ return this.recipientDatabase.setNickname(recipient, nickname);
- return recipientDatabase.setNickname(recipient, nickname);
```

?

veken1199 replied:
Yes, your approach is better in terms of readability

g-harel commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,70 @@
+    public boolean removeNickname(Recipient recipient) {
+        setupDatabaseHandler();
+
+        // TODO call the correct remove method
```
This means we cannot merge yet?

veken1199 replied:
We can merge, actuall @Dan-Yoo  needs to continue working from this code from his branch

g-harel replied:
So remove this from your branch then ?

veken1199 replied:
I think it is nice to keep it to guide him and he ll know what is missing and from where he wants to start. But maybe we can rename this branch to "nickname_backend" and will not be merged until everything is finished.. @Dan-Yoo what do you think ?

g-harel commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,70 @@
+import org.thoughtcrime.securesms.database.DatabaseFactory;
+import org.thoughtcrime.securesms.database.RecipientDatabase;
+import org.thoughtcrime.securesms.recipients.Recipient;
+
```
extra line

veken1199 replied:
noted

g-harel replied:
It's still there

g-harel commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,70 @@
+     * @return boolean indicating whether the nickname was set or not
+     */
+    public boolean setNickname(Recipient recipient, String nickname) {
+        setupDatabaseHandler();
```
this. ?

veken1199 replied:
sure

g-harel commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    /**
+     * This method responsible for setting the nick name
```
Same comment about the first sentence.

veken1199 replied:
again, since it is javadoc, I was thinking to write something that is very clear and explains the reason for this methods existence

g-harel replied:
it is clear without the need of the first sentence

g-harel commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,70 @@
+
+    // TODO by DAN
+    public boolean removeNickname(Recipient recipient) {
+        setupDatabaseHandler();
```
this. ?

g-harel commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,70 @@
+import org.thoughtcrime.securesms.recipients.Recipient;
+
+
+public class NicknameHandler {
```
I don't understand the point of this whole class. It seems to be some kind of database helper, but right now it does almost nothing.

veken1199 replied:
this method allows us to write tests and check if nickname feature works, if we dont do this, fragment will right away call databases and we can not test fragments

g-harel replied:
But couldn't the "setDatabase" method be on the fragment?

g-harel commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,70 @@
+     * in production only.
+     * @return NicknameHandler
+     */
+    public NicknameHandler setupDatabaseHandler() {
```
If the other methods call this, does it need to be public? Would make using the class a bit easier.

veken1199 replied:
I dont understand

g-harel replied:
`setupDatabaseHandler` is being called at the start of all the other methods. This means (in my understanding) that it does not need to be called from outside the `NicknameHandler` class and can be made private.

g-harel commented on src/org/thoughtcrime/securesms/database/DatabaseFactory.java:
```diff
@@ -116,7 +116,8 @@
   private static final int MORE_RECIPIENT_FIELDS                           = 47;
   private static final int INTRODUCED_PINNED_MESSAGES                      = 48;
-  private static final int DATABASE_VERSION                                = 48;
+  private static final int INTRODUCED_NICK_NAMES                           = 49;
```
nickname is one word

veken1199 replied:
sure

g-harel commented on src/org/thoughtcrime/securesms/database/DatabaseFactory.java:
```diff
@@ -1416,11 +1417,15 @@ public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
       }
 
+      if (oldVersion < INTRODUCED_NICK_NAMES) {
+        db.execSQL("ALTER TABLE recipient_preferences ADD COLUMN nick_name TEXT DEFAULT 'null';");
```
Is this the `"null"` string or `NULL` ?

veken1199 replied:
I had a hard time setting db attr to NULL using update statement since they always set it to "NULL" so i decided to make the default "null"  since java knows null but not NULL for some reason,

g-harel commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -50,11 +50,12 @@
   private static final String SIGNAL_PROFILE_NAME     = "signal_profile_name";
   private static final String SIGNAL_PROFILE_AVATAR   = "signal_profile_avatar";
   private static final String PROFILE_SHARING         = "profile_sharing_approval";
+  private static final String NICK_NAME               = "nick_name";
```
nickname

veken1199 replied:
sure

veken1199 replied:
but the db attribute is set to nick_name, is that ok ?

g-harel commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -171,6 +174,7 @@ public BlockedReader readerForBlocked(Cursor cursor) {
     String  signalProfileAvatar   = cursor.getString(cursor.getColumnIndexOrThrow(SIGNAL_PROFILE_AVATAR));
     boolean profileSharing        = cursor.getInt(cursor.getColumnIndexOrThrow(PROFILE_SHARING))      == 1;
 
+
```
.

veken1199 replied:
sure

g-harel commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -197,7 +201,8 @@ public BlockedReader readerForBlocked(Cursor cursor) {
                                              systemPhoneLabel, systemContactUri,
-                                             signalProfileName, signalProfileAvatar, profileSharing));
+                                             signalProfileName, signalProfileAvatar, profileSharing,
+                                              nickName));
```
Is there a line limit in the coding style guide?

veken1199 replied:
no

g-harel commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -236,6 +241,12 @@ public void setBlocked(@NonNull Recipient recipient, boolean blocked) {
+  public boolean setNickname(@Nullable Recipient recipient, String nickname) {
+    ContentValues values = new ContentValues();
+    values.put(NICK_NAME, nickname);
+    return processNicknameSqlRequest(values, recipient.getAddress()) > 0 ? true : false;
```
lol `true ? true : false;`

veken1199 replied:
where do you read the first true?

g-harel commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -314,6 +325,14 @@ public void setProfileSharing(@NonNull Recipient recipient, @SuppressWarnings("S
 
+  private int processNicknameSqlRequest(ContentValues values, Address address) {
+    SQLiteDatabase db = databaseHelper.getWritableDatabase();
+    return db.update(this.TABLE_NAME, values, ADDRESS + " = ? AND " +
```
I feel like this method call can be formatted a bit nicer.

veken1199 replied:
can you propose something ??? this method will also be used by removeNickname, and the logic that it implements is that if the new nickname is the same as the old nickname then return false. otherwise return true

g-harel replied:
Maybe something like this. Just to make it a bit clearer.

```java
return db.update(
    this.TABLE_NAME,
    values,
        ADDRESS + " = ? AND " +
        NICKNAME + " NOT Like ? ",
    new String[] {
        address.serialize(),
        values.getAsString(NICKNAME)
    });
```

g-harel commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -440,6 +459,7 @@ public void finish() {
     private final int             expireMessages;
     private final RegisteredState registered;
     private final byte[]          profileKey;
+    private final String          nickName;
```
nickname

veken1199 replied:
sure

g-harel commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -448,6 +468,7 @@ public void finish() {
     private final String          signalProfileAvatar;
     private final boolean         profileSharing;
 
+
```
.

g-harel replied:
still there

g-harel commented on test/unitTest/java/org/thoughtcrime/securesms/NicknameMocks.java:
```diff
@@ -0,0 +1,61 @@
+import static org.mockito.Mockito.mockingDetails;
+import static org.mockito.Mockito.when;
+
+
```
.

g-harel replied:
Still there

g-harel commented on test/unitTest/java/org/thoughtcrime/securesms/database/NicknameTests.java:
```diff
@@ -0,0 +1,47 @@
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+
```
.

g-harel commented on test/unitTest/java/org/thoughtcrime/securesms/database/NicknameTests.java:
```diff
@@ -0,0 +1,47 @@
+        super.setupRemoveNicknameMethod();
+        super.setupStaticRecipientDatabase();
+        super.setupSetNicknameMethod();
+
```
.

g-harel commented on test/unitTest/java/org/thoughtcrime/securesms/database/NicknameTests.java:
```diff
@@ -0,0 +1,47 @@
+        verify(recipientDatabase, times(2)).setNickname(super.recipient, "test");
+        verify(recipientDatabase, times(2)).setNickname(super.recipient, "new test");
+    }
+
```
O_O

veken1199 replied:
magicccccc

g-harel commented on test/unitTest/java/org/thoughtcrime/securesms/NicknameMocks.java:
```diff
@@ -0,0 +1,61 @@
+
+    }
+
+    // TODO add removeNickname mock
```
should do this before merge

veken1199 replied:
no, we need to merge, so that @Dan-Yoo starts working on his "remove" feature and write his own tests

g-harel replied:
So remove it from here ?

veken1199 replied:
we will need to talk to @Dan-Yoo first,

veken1199 replied:
if we remove it @Dan-Yoo will spend time trying to understand and re designing the wheel. in this case, he will right away go to the solution.

g-harel replied:
... Or we can communicate and he will know what to do ?

g-harel commented on test/unitTest/java/org/thoughtcrime/securesms/NicknameMocks.java:
```diff
@@ -0,0 +1,61 @@
+public class NicknameMocks extends BaseUnitTest {
+    protected DatabaseFactory   databaseFactory;
+    protected Recipient         recipient;
+    protected Recipient         recipient2;
```
this one is not used anywhere.

veken1199 replied:
what do you mean?
they are, maybe not recipient2

g-harel replied:
`recipient2` is not used anywhere

g-harel commented on test/unitTest/java/org/thoughtcrime/securesms/nickname/NicknameHandlerTests.java:
```diff
@@ -0,0 +1,82 @@
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+
```
.

veken1199 replied:
ok

g-harel commented on test/unitTest/java/org/thoughtcrime/securesms/nickname/NicknameHandlerTests.java:
```diff
@@ -0,0 +1,82 @@
+
+        verify(recipientDatabase, times(2)).setNickname(recipient,"test");
+        verify(recipientDatabase, times(2)).setNickname(recipient,"new test");
+
```
...............................................................................................................................................................

veken1199 replied:
????

g-harel replied:
extra lines again

cool-appz committed:
# related to #38 : General code formatting after the first review

g-harel commented on test/unitTest/java/org/thoughtcrime/securesms/nickname/NicknameHandlerTests.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    @Test
+    public void testSetNicknameSameOldeName() {
```
...OldName*

veken1199 replied:
aghh

g-harel reviewed (dismissed):
What the heck was that last commit, you changed almost 2k lines !?

There are also some un-addressed comments remaining

cool-appz committed:
# related to #38 : General code formatting after the Second review

cool-appz committed:
# related to #38 : General code formatting after the Second review - reverted DatabaseFactory Format

cool-appz committed:
# related to #38 : General code formatting after the Second review - reverted RecipientDatabase Format

g-harel commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -117,7 +118,8 @@ public static RegisteredState fromId(int id) {
           SIGNAL_PROFILE_AVATAR + " TEXT DEFAULT NULL, " +
-          PROFILE_SHARING + " INTEGER DEFAULT 0);";
+          PROFILE_SHARING + " INTEGER DEFAULT 0, " +
+              NICKNAME + " TEXT DEFAULT 'null');";
```
Is the extra space deliberate ?

g-harel commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -236,6 +241,12 @@ public void setBlocked(@NonNull Recipient recipient, boolean blocked) {
+  public boolean setNickname(@Nullable Recipient recipient, String nickname) {
+    ContentValues values = new ContentValues();
+    values.put(NICKNAME, nickname);
+    return processNicknameSqlRequest(values, recipient.getAddress()) > 0 ? true : false;
```
`>` already evaluates to a boolean

erinb90 commented on src/org/thoughtcrime/securesms/NicknameHandler.java:
```diff
@@ -0,0 +1,69 @@
+import org.thoughtcrime.securesms.recipients.Recipient;
+
+public class NicknameHandler {
+    private Context context;
```
shouldn't it be formatted like:
```
private Context           context;
private RecipientDatabase recipientDatabase;
```

erinb90 commented on src/org/thoughtcrime/securesms/database/RecipientDatabase.java:
```diff
@@ -482,6 +503,7 @@ public void finish() {
       this.signalProfileName     = signalProfileName;
       this.signalProfileAvatar   = signalProfileAvatar;
       this.profileSharing        = profileSharing;
+      this.nickname = nickname;
```
needs the spaces before the =

cool-appz committed:
# related to #38 : General code formatting after the Third review
1- Removed TODO parts related for removeNickname functionality
2- Fixed a functional inconsistency in PinnedMessagesMocks
3- General code formatting

cool-appz committed:
# related to #38 : General code formatting after the Third review
1- General code formatting

g-harel committed:
#38 formatting changes

erinb90 reviewed (approved):
👍
